File: solutions/_1721_swapping_nodes_in_a_linked_list/Solution.java

package _1721_swapping_nodes_in_a_linked_list;

class Solution {
    public ListNode swapNodes(ListNode head, int k) {
        ListNode cur = head;
        for (int i = 1; i < k; i++) {
            cur = cur.next;
        }

        ListNode left = cur;
        ListNode right = head;
        while (cur.next != null) {
            cur = cur.next;
            right = right.next;
        }

        int tmp = left.val;
        left.val = right.val;
        right.val = tmp;

        return head;
    }
}

/*
 * https://leetcode.com/problems/swapping-nodes-in-a-linked-list/
 */

File: solutions/_1721_swapping_nodes_in_a_linked_list/solution.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        # 1. Find the kth node from the start
        cur = head
        for _ in range(1, k):
            cur = cur.next
        left = cur

        # 2. Find the kth node from the end
        right = head
        while cur.next:
            cur = cur.next
            right = right.next

        # 3. Swap the values of the kth node from the start and the kth node from the end
        left.val, right.val = right.val, left.val
        return head


# https://leetcode.com/problems/swapping-nodes-in-a-linked-list/
